### PR TITLE
fix/PSD-3947-SCPN-url_generation_error

### PR DESCRIPTION
**Pull request by mirotriad**
# Fix component routing issue and refactor tests

## Description
This PR addresses a production error where the notification component build path was failing due to missing component IDs. It also includes a comprehensive refactor of the associated tests to improve maintainability and follow best practices.

## Changes
- Fixed component routing in `ProductController` by ensuring component creation with `first_or_create!`
- Refactored tests to follow single expectation pattern
- Added proper test coverage for all notification states (nanomaterials, multi-component, single component)
- Updated multi-component tests to use real data setup instead of stubs

## Related Issues
Fixes Sentry: `ActionController::UrlGenerationError` in `ResponsiblePersons::Notifications::ProductController#show`

## Testing
- All existing functionality remains unchanged
- Added new test cases to cover edge cases
- All tests passing
- RuboCop compliant

### CODE REVIEW

File: cosmetics-web/app/controllers/responsible_persons/notifications/product_controller.rb [hunk 1]
```diff
@@ -26,7 +26,8 @@ def show
     case step
     when :completed
       @notification.set_state_on_product_wizard_completed!
-      render template: "responsible_persons/notifications/task_completed", locals: { continue_path: }
+      @continue_path = continue_path
```
Comment (alan-at-work) [inline cosmetics-web/app/controllers/responsible_persons/notifications/product_controller.rb:29]: I don't think `@continue_path` is needed in this case. The local variable `continue_path` is already available. And this is not passed to the view ;)

File: cosmetics-web/app/controllers/responsible_persons/notifications/product_controller.rb [hunk 2]
```diff
@@ -26,7 +26,8 @@ def show
...
+      render template: "responsible_persons/notifications/task_completed", locals: { continue_path: continue_path }
```
Comment (tahir-khalid) [inline cosmetics-web/app/controllers/responsible_persons/notifications/product_controller.rb:30]: The template path can this be a variable or a helper function to return that value from a constant if need be.

Comment (tahir-khalid) [inline cosmetics-web/app/controllers/responsible_persons/notifications/product_controller.rb:30]: See previous comment could be something `def responsible_persons_notif_task_completed_path...end`

File: cosmetics-web/spec/controllers/responsible_persons/notifications/product_controller_spec.rb [hunk 3]
```diff
@@ -52,12 +52,88 @@
...
+      context "when notification has nanomaterials" do
+        let(:nano_material) { create(:nano_material, notification: notification) }
+
+        before do
+          notification.nano_materials << nano_material
+        end
+
+        it "renders task completed template" do
+          get(:show, params: params_with_completed)
+          expect(response).to render_template("responsible_persons/notifications/task_completed")
```
Comment (tahir-khalid) [inline cosmetics-web/spec/controllers/responsible_persons/notifications/product_controller_spec.rb:75]: Can this be returned from a private function (see previous comment for example) or a let so its not hardcoded and better readability thanks.

File: cosmetics-web/spec/controllers/responsible_persons/notifications/product_controller_spec.rb [hunk 4]
```diff
@@ -52,12 +52,88 @@
...
+      context "when notification is multi-component" do
+        before do
+          notification.components.destroy_all
+          notification.update_column(:state, "ready_for_components")
+          create_list(:component, 2, notification: notification)
+        end
+
+        it "renders task completed template" do
+          get(:show, params: params_with_completed)
+          expect(response).to render_template("responsible_persons/notifications/task_completed")
```
Comment (tahir-khalid) [inline cosmetics-web/spec/controllers/responsible_persons/notifications/product_controller_spec.rb:93]: Same as above can you apply that refactoring suggestion to the rest if you decide to go that path thanks.

File: cosmetics-web/app/controllers/responsible_persons/notifications/product_controller.rb (hunk 1 above)
Comment (mirotriad) [reply cosmetics-web/app/controllers/responsible_persons/notifications/product_controller.rb:29]: continue_path was being passed as a local variable to the template but not being assigned as an instance variable, so assigns(:continue_path) was returning nil

File: cosmetics-web/app/controllers/responsible_persons/notifications/product_controller.rb (hunk 2 above)
Comment (mirotriad) [reply cosmetics-web/app/controllers/responsible_persons/notifications/product_controller.rb:30]: I think this is not needed - the path is used only once, so for each render, for example, if there are 5 paths, we will create 5 methods, which is an overhead don't you think?

**Review by alan-at-work — APPROVED** [review]

File: cosmetics-web/spec/controllers/responsible_persons/notifications/product_controller_spec.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (mirotriad) [reply cosmetics-web/spec/controllers/responsible_persons/notifications/product_controller_spec.rb:75]: Thanks for the suggestion @alan-at-work! I've split up the specs and added the continue_path assertions. This should give us better test coverage of the navigation flow while keeping the specs clean and focused. Let me know if you'd like any adjustments to this approach.

**Review by alan-at-work — APPROVED** [review]
